libs/libbuiltin/libgcc/gcov.c

```diff
@@ -0,0 +1,290 @@
+
+#include <gcov.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define GCOV_DATA_MAGIC       (0x67636461)
+#define GCOV_NOTE_MAGIC       (0x67636e6f)
+#define GCOV_FILENAME_MAGIC   (0x6763666e)
+
+#define GCOV_TAG_FUNCTION     (0x01000000)
+#define GCOV_TAG_COUNTER_BASE (0x01a10000)
+
+#define GCOV_TAG_FOR_COUNTER(count) \
+  (GCOV_TAG_COUNTER_BASE + ((uint32_t)(count) << 17))
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+typedef unsigned int gcov_unsigned_t;
+
+/* Information about counters for a single function
+ *
+ * This data is generated by gcc during compilation and doesn't change
+ * at run-time.
+ */
+
+struct gcov_ctr_info
+{
+  unsigned int num;      /* Number of counter values for this type */
+  FAR gcov_type *values; /* Array of counter values for this type */
+};
+
+/* Profiling meta data per function
+ *
+ * This data is generated by gcc during compilation and doesn't change
+ * at run-time.
+ */
+
+struct gcov_fn_info
+{
+  FAR const struct gcov_info *key; /* Comdat key */
+  unsigned int ident;              /* Unique ident of function */
+  unsigned int lineno_checksum;    /* Function lineno checksum */
+  unsigned int cfg_checksum;       /* Function cfg checksum */
+  struct gcov_ctr_info ctrs[1];    /* Instrumented counters */
+};
+
+/****************************************************************************
+ * Private Functions Prototypes
+ ****************************************************************************/
+
+static size_t gcov_convert(FAR uint8_t *buffer,
+                           FAR const struct gcov_info *info);
+static size_t gcov_buff_size(FAR const struct gcov_info *info);
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+FAR struct gcov_info *__gcov_info_start;
+FAR struct gcov_info *__gcov_info_end;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void dump_counter(FAR void *buffer, FAR size_t *off, uint64_t v)
+{
+  memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
+  *off += sizeof(uint64_t);
+}
+
+static void dump_unsigned(FAR void *buffer, FAR size_t *off, uint32_t v)
+{
+  memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
+  *off += sizeof(uint32_t);
+}
+
+static size_t gcov_buff_size(FAR const struct gcov_info *info)
+{
+  uint32_t iter;
+  uint32_t iter_1;
+  uint32_t iter_2;
+
+#ifdef GCOV_12_FORMAT
+  uint32_t size = sizeof(uint32_t) * 4;
```

Comment:
why 4 or 3? please add some comments or define to macro

Reply:
This is the calculation buffer size according to the gcda file format.

The gcov_buff_size function has been deleted. We can get the buffer size through the gcov_convert function.
The meaning and size of each field is easy to see in the gcov_convert function